### PR TITLE
Support NixOS

### PR DESCRIPTION
**Pull request by dmadisetti**
I don't have the time to package this right now, but I thought I would share the minor changes needed to get this working for nix

required packages `nix-shell -p grub2_efi mtools OVMF python38Packages.pip qemu xorriso`

execute `./grub2-theme-preview /boot/theme`

the best approach is probably to allow for more flags (or pull from environmental variables)

### CODE REVIEW

**Comment by hartwork** [comment]
@dmadisetti the path patched in looks like it will change with at least every release. Then patching this path in should probably be done in NixOS packaging rather than in here. What do you think?

**Comment by dmadisetti** [comment]
@hartwork It's even worse! This will break every content change. That garbage in front of the path is generated from hashing contents and dependency (not sure to what degree). Nix prides it's self on reproducibility, but it can be a pain.

I think the best way to do this is as follows:

```python
import os

def _grub2_directory(platform):
    return  "%s/%s" % (os.environ.get("GRUB2_PREVIEW_ROOT", "/usr/lib/grub"), platform)

# ...
declared_candidate = os.environ.get("GRUB2_PREVIEW_OVMF_CANDIDATE")
if declared_candidate:
   candidates.append(declared_candidate)
```

This is potentially a feature for other systems.
It's pretty typical of nix to wrap programs. So we would do that here. Preferred to passing in flags, since multiple arguments might lead to undefined behavior, and patching can be pretty brittle.

**Comment by dmadisetti** [comment]
So this works really nicely. Example: `nix run ".#" -- /boot/theme --resolution 1920x1080`. Once published `nix run "github:hartwork/grub2-theme-preview#"`.

image

Still keeping this as a draft, because I had to unpin requirements.... Not sure how to get around that unless [mach](https://github.com/DavHau/mach-nix/tree/master) catches up to pypi....

File: grub2_theme_preview/__main__.py [hunk 1]
```diff
@@ -366,7 +366,9 @@ def _grub2_ovmf_tuple():
         '/usr/share/edk2-ovmf/x64/OVMF_CODE.fd',  # Arch Linux and its derivatives
         '/usr/share/OVMF/OVMF_CODE.fd',  # Debian and its derivatives
         '/usr/share/edk2/ovmf/OVMF_CODE.fd',  # Fedora (and its derivatives?)
+        os.environ.get('GRUB2_PREVIEW_OVMF_CANDIDATE', ''), # NixOS and potentially others
     ]
+
```
Comment (hartwork) [inline grub2_theme_preview/__main__.py:371]: Let's adjust the variable name a bit and only check that very file if given, i.e.:
```python
omvf_image = os.environ.get('G2TP_OVMF_IMAGE')
if omvf_image is not None:
    candidates = [omvf_image]
else:
    candidates = [
        '/usr/share/edk2-ovmf/OVMF_CODE.fd',  # Gentoo and its derivatives
        '/usr/share/edk2-ovmf/x64/OVMF_CODE.fd',  # Arch Linux and its derivatives
        '/usr/share/OVMF/OVMF_CODE.fd',  # Debian and its derivatives
        '/usr/share/edk2/ovmf/OVMF_CODE.fd',  # Fedora (and its derivatives?)
    ]
```
What do you think?

File: grub2_theme_preview/__main__.py [hunk 2]
```diff
@@ -339,7 +339,7 @@ def parse_command_line(argv):
 
 
 def _grub2_directory(platform):
-    return '/usr/lib/grub/%s' % platform
+    return  '%s/%s' % (os.environ.get('GRUB2_PREVIEW_ROOT', '/usr/lib/grub'), platform)
```
Comment (hartwork) [inline grub2_theme_preview/__main__.py:342]: Let's use name `G2TP_GRUB_LIB` for the variable, please.

File: grub2_theme_preview/__main__.py [hunk 3]
```diff
@@ -144,9 +144,9 @@ def _make_grub_cfg_load_our_theme(grub_cfg_content, source_type, resolution_or_n
     epilog_chunks = [
         # Ensure that we always have one or more menu entries
         '',
-        'submenu \'Reboot / Shutdown\' {',
-        '    menuentry Reboot { reboot }',
-        '    menuentry Shutdown { halt }',
+        'submenu \'Reboot / Shutdown\' --class shutdown {',
+        '    menuentry Reboot --class restart { reboot }',
+        '    menuentry Shutdown --class shutdown { halt }',
```
Comment (hartwork) [inline grub2_theme_preview/__main__.py:149]: If this has practical value, please move it to a dedicated pull request and elaborate on it's practical effect. Thank you!

**Review by hartwork — CHANGES_REQUESTED** [review]
Hi @dmadisetti we can add the envornment variables but the NixOS packaging and the requirements unpins have no place in the upstream repository, please revert them. On a sidenote, `requirements.txt` is only used in CI and does not affect the PyPI package where dependencies are not pinned to play well with neighbor packages. The pinning helps CI stability. Please see my comments on details below:

File: grub2_theme_preview/__main__.py (hunk 1 above)
Comment (dmadisetti) [reply grub2_theme_preview/__main__.py]: LGTM

**Comment by dmadisetti** [comment]
Thanks for the review @hartwork 

🤔 The only dep that is actually needed was `setuptools`, but I also didn't implement testing with the nix package.

Changing from draft status, since I think this is actually acceptable now (no need to play with `requirements.txt`!)

**Review by hartwork — CHANGES_REQUESTED** [review]
Hi @dmadisetti thanks for the adjustments! Could you drop files `flake.nix` and `flake.lock` from the pull request, please?

**Comment by dmadisetti** [comment]
😕 nix really isn't supported without them (at least flake.nix). user repositories are not searchable (and not super used tbh), and nixpkgs can be a pain to merge into- but if you add a license file I'm happy to maintain a fork. I also am asking around the community about the best way to do this otherwise

**Comment by hartwork** [comment]
@dmadisetti I'm not sure I understand. Another tool of mine — [git-delete-merged-branches](https://github.com/hartwork/git-delete-merged-branches) — is packaged in NixOS at https://github.com/hartwork/nixpkgs/blob/master/pkgs/applications/version-management/git-and-tools/git-delete-merged-branches/default.nix and I would imagine that that same approach would work for packaging grub2-theme-preview. What do you think?

**Comment by hartwork** [comment]
PS: I'd be happy to merge the environment variable part in any case, if the pull request can be made to be about just that.

**Comment by hartwork** [comment]
PS: @dmadisetti I can make a new release off `master` with the environment variable patch merged for packaging in NixOS on rather short notice.

**Comment by dmadisetti** [comment]
Removed the flake definitions, I will work on nixos support later

Sorry I didn't follow up yesterday- I was still trying to figure out the best way to package this for a broader audience. I will try to merge my flake definitions into nixpkgs, but I have not had the fastest turn around experiences

**Review by hartwork — CHANGES_REQUESTED** [review]
@dmadisetti thanks for the adjustments! The CI is currently red for two small things introduced here, could you fix those? The detailed log is at https://github.com/hartwork/grub2-theme-preview/runs/5956631013?check_suite_focus=true#step:6:23 . Thanks!

**Comment by hartwork** [comment]
@dmadisetti PS: The changes needed to make CI happy are these:

```diff
--- return  '%s/%s'
+++ return '%s/%s'
```

```diff
--- if omvf_image is not None: # Support non-standard locations e.g. NixOS
+++ if omvf_image is not None:  # Support non-standard locations e.g. NixOS
```

**Comment by hartwork** [comment]
@dmadisetti the last thing the CI is complaining about now will probably come as a bit of a surprise. This patch is needed:
```diff
diff --git a/grub2_theme_preview/__main__.py b/grub2_theme_preview/__main__.py
index d8fd05b..611965a 100644
--- a/grub2_theme_preview/__main__.py
+++ b/grub2_theme_preview/__main__.py
@@ -339,7 +339,7 @@ def parse_command_line(argv):
 
 
 def _grub2_directory(platform):
-    return '%s/%s' % (os.environ.get('G2TP_GRUB_LIB', '/usr/lib/grub'), platform)
+    return '{}/{}'.format(os.environ.get('G2TP_GRUB_LIB', '/usr/lib/grub'), platform)
 
 
 def _grub2_platform():
```
Not sure why pyupgrade was happy with the old code then. Thanks for your patience!

**Comment by dmadisetti** [comment]
Hm, strange. That pattern is elsewhere

**Review by hartwork — APPROVED** [review]
> Hm, strange. That pattern is elsewhere

@dmadisetti yes; I believe the tool is not yet detecting all flavors of the pattern. Thanks for the adjustments and the pull request! :+1:

**Comment by hartwork** [comment]
@dmadisetti 2.6.0 now released, including PyPI
